In [1]:
import pandas as pd
import numpy as np
import scipy.stats # for binomial

1. Using `pandas`, read the `csv` file `ita_2000.csv` directly from the internet.

In [2]:
url = 'https://raw.githubusercontent.com/StefanoAllesina/namepairs/master/data/ita_2000.csv'
data = pd.read_csv(url)

2. Take a look at the data: for each of the 52,000 Italian professors who were active in 2000, it reports a code for the first name, a code for the last name, the gender, rank, university and discipline. 

In [3]:
data

Unnamed: 0  first_id  last_id gender                 rank  \
0               1      5001        3      F  associate professor   
1               2      2749       23      M  assistant professor   
2               3      6516       23      M            professor   
3               4      8635       94      M  associate professor   
4               5      3064      588      M            professor   
5               6      6703      685      F  associate professor   
6               7      7117      716      M  associate professor   
7               8      3938     1232      M            professor   
8               9      5403     2059      F  assistant professor   
9              10      2946     2103      F  associate professor   
10             11      6516     2219      M  associate professor   
11             12      2897     3579      M  assistant professor   
12             13      6223     3813      M            professor   
13             14      5113     3947      M  associate professor   
14             15       706     5379      F  assistant professor   
15             16      4343     5379      M  associate professor   
16             17      7782     5633      M  associate professor   
17             18      1227     6219      M  associate professor   
18             19      5907     7791      M            professor   
19             20       434     8350      M  associate professor   
20             21      5414     8517      F            professor   
21             22      6285     8708      M  associate professor   
22             23      5288     9697      M  assistant professor   
23             24      2946    10378      F            professor   
24             25      6566     9792      F  assistant professor   
25             26      1358     9930      M  associate professor   
26             27      1691    10652      M            professor   
27             28      3766    10794      M            professor   
28             29      5114    11082      F  assistant professor   
29             30      6682    11120      M            professor   
...           ...       ...      ...    ...                  ...   
51974       51975       250     9417      M            professor   
51975       51976      1916    10609      F  associate professor   
51976       51977      3673    10751      M  associate professor   
51977       51978      2783    10787      M            professor   
51978       51979      3766    12525      M            professor   
51979       51980      3153    12687      M            professor   
51980       51981      2749    14704      M            professor   
51981       51982       878    18820      M            professor   
51982       51983      2749    21179      M            professor   
51983       51984      3329    21579      M            professor   
51984       51985      5054    21990      M  associate professor   
51985       51986      3961    24460      M  associate professor   
51986       51987      3938    26326      M            professor   
51987       51988      6840    26968      M            professor   
51988       51989      1402    27540      M            professor   
51989       51990      5907    28014      M            professor   
51990       51991      7597    30858      M            professor   
51991       51992      5113     8078      M            professor   
51992       51993      7323    10220      M  associate professor   
51993       51994      7924    17784      M            professor   
51994       51995      5156    22830      M            professor   
51995       51996      7568    23319      M  assistant professor   
51996       51997      8617    32229      M            professor   
51997       51998      3064    16831      M            professor   
51998       51999      4147    21892      M            professor   
51999       52000      7568     4757      M            professor   
52000       52001       811    11534      F  assistant professo

3. Take for example `Eng-Ind` (industrial engineering): there are 396 women and 3734 men (for a total of 4130 professors). Are there too few women? To test this, compute the probability of sampling less than 396 women out of 4130 professors when drawing them at random without repetition. Contrast this with what expected using the binomial distribution. 

We're going to write a function, because we need to repeat this for all disciplines below

In [4]:
def prob_women(data, discipline = "Eng-Ind", nrep = 1000):
    # count number of women in discipline
    data_disc = data[data['sector'] == discipline]
    observed = sum(data_disc['gender'] == 'F')
    total = data_disc.shape[0]
    prob = 0.0
    for i in range(nrep):
        sampled = sum(data.sample(n = total)['gender'] == 'F')
        if sampled <= observed:
            prob = prob + 1.0
    prob = prob / nrep
    # now compute probability according to binomial
    # q is the proportion of women
    q = sum(data['gender'] == 'F') / data.shape[0]
    # this is the exact probability
    binom_prob = scipy.stats.binom.cdf(observed, total, q)
    return {"discipline": discipline, 
            "F": observed,
            "tot": total,
            "prob": round(prob, 3),
            "binom_prob": round(binom_prob, 3)
            }

In [5]:
prob_women(data, discipline = "Eng-Ind", nrep = 1500)

{'discipline': 'Eng-Ind',
 'F': 396,
 'tot': 4130,
 'prob': 0.0,
 'binom_prob': 0.0}

4. Repeat for each discipline: which ones have an "excess" of women? Which ones a "scarcity"?

In [6]:
# extract all disciplines
disciplines = set(data['sector'])

In [7]:
for d in disciplines:
    tmp = prob_women(data, d, 2500)
    print("===========")
    for key, val in tmp.items():
        print(key, val)
    print("===========")

discipline Econ
F 925
tot 3493
prob 0.0
binom_prob 0.001
discipline Bio
F 1949
tot 4530
prob 1.0
binom_prob 1.0
discipline Phys
F 368
tot 2408
prob 0.0
binom_prob 0.0
discipline Math
F 994
tot 2926
prob 1.0
binom_prob 1.0
discipline Hist-Ped-Psi
F 1630
tot 4164
prob 1.0
binom_prob 1.0
discipline Soc
F 384
tot 1287
prob 0.8
binom_prob 0.79
discipline Agr
F 702
tot 2787
prob 0.0
binom_prob 0.0
discipline Eng-Civ
F 635
tot 3368
prob 0.0
binom_prob 0.0
discipline Eng-Ind
F 396
tot 4130
prob 0.0
binom_prob 0.0
discipline Hum
F 2762
tot 5215
prob 1.0
binom_prob 1.0
discipline Med
F 1996
tot 9559
prob 0.0
binom_prob 0.0
discipline Geo
F 285
tot 1276
prob 0.0
binom_prob 0.0
discipline Law
F 1008
tot 3721
prob 0.008
binom_prob 0.009
discipline Chem
F 975
tot 3140
prob 0.997
binom_prob 0.997


6. Now turn to last names: a scarcity of last names in a discipline could signal nepotistic practices (as children would bear the same last name as the father). Take `Law`: there are 3721 professors, and we find 3037 names. Are these many names, or less than we would expect? Compute an approximate p-value for the probability of finding fewer names when sampling from the set of professors. 

Again, we are going to write a function, so that it is easy to run this for each discipline.

In [8]:
def prob_last_names(data, discipline = "Law", nrep = 1500):
    data_disc = data[data['sector'] == discipline]
    observed = len(set(data_disc['last_id']))
    total = data_disc.shape[0]
    prob = 0.0
    for i in range(nrep):
        sampled = len(set(data.sample(n = total)['last_id']))
        if sampled <= observed:
            prob = prob + 1.0
    prob = prob / nrep
    return {"discipline": discipline, 
            "observed": observed,
            "tot": total,
            "prob": round(prob, 3)
            }

In [9]:
prob_last_names(data)

{'discipline': 'Law', 'observed': 3037, 'tot': 3721, 'prob': 0.0}

7. Repeat for each discipline: which ones are most likely to be impacted by nepotism?

In [10]:
for d in disciplines:
    tmp = prob_last_names(data, d, 2500)
    print("===========")
    for key, val in tmp.items():
        print(key, val)
    print("===========")

discipline Econ
observed 2961
tot 3493
prob 0.031
discipline Bio
observed 3712
tot 4530
prob 0.013
discipline Phys
observed 2125
tot 2408
prob 0.078
discipline Math
observed 2522
tot 2926
prob 0.017
discipline Hist-Ped-Psi
observed 3506
tot 4164
prob 0.624
discipline Soc
observed 1196
tot 1287
prob 0.284
discipline Agr
observed 2412
tot 2787
prob 0.016
discipline Eng-Civ
observed 2862
tot 3368
prob 0.021
discipline Eng-Ind
observed 3399
tot 4130
prob 0.002
discipline Hum
observed 4335
tot 5215
prob 1.0
discipline Med
observed 6718
tot 9559
prob 0.0
discipline Geo
observed 1179
tot 1276
prob 0.088
discipline Law
observed 3037
tot 3721
prob 0.0
discipline Chem
observed 2683
tot 3140
prob 0.012
